Allow "modern C" options when compiling

rkta opened the pull request:
This PR enables building with -Werror=implicit-function-declaration,
-Werror=implicit-int and -Werror=strict-prototypes.

It is related to https://bugs.gentoo.org/show_bug.cgi?id=870412 and
https://bugzilla.redhat.com/show_bug.cgi?id=2142177

rkta committed:
Enable building with "modern C" build options

Enable building with "modern C" build options

See the relevant Gentoo[0] and RedHat[1] bugs.

[0]: https://bugs.gentoo.org/show_bug.cgi?id=870412 an
[1]: https://bugzilla.redhat.com/show_bug.cgi?id=2142177

This enable to use
make OPTS="-Werror=implicit-function-declaration -Werror=implicit-int
-Werror=strict-prototypes"

rkta committed:
Fix build error with --disable-color

With --disable-color highIntensityColors is not declared.

rkta commented:
@N-R-K This might be relevant for you.

N-R-K commented on w3mhelperpanel.c:
```diff
@@ -167,7 +167,7 @@ int
 {
     Str mailcapfile;
-    extern char *getenv();
+    extern char *getenv(const char *);
```
why are we redeclaring things that should be declared in standard headers already :)

N-R-K commented on istream.h:
```diff
@@ -75,8 +75,8 @@ struct ssl_stream {
-    int (*read) ();
-    void (*close) ();
+    int (*read) (void *, unsigned char *, int);
+    void (*close) (void *, int, int *);
```
These functions are a bit problematic because they can either be stdio functions or unix functions that operat on fd. Specifying `FILE *` gives the wrong impression that they're meant for stdio only.

Personally I'd go with either void parameter as the type `f(void)` or just a regular `void *` and then cast it to proper type on usage. A union might also be a possibility.

N-R-K reviewed (commented):
I remember fixing a bunch of them. But I don't think I caught them all due to how many build configuration is possible.

Your patch mostly looks fine, but I have a few nitpicks.

rkta committed:
Remove declaration of libc functions

rkta commented:
On Tue, Dec 19, 2023 at 04:22:00AM -0800, NRK wrote:
> @N-R-K commented on this pull request.
[...]
> > @@ -167,7 +167,7 @@ int
>  main(void)
>  {
>      Str mailcapfile;
> -    extern char *getenv();
> +    extern char *getenv(const char *);
> 
> why are we redeclaring things that should be declared in standard
> headers already :)

Oh, it's an old tradition - we are doing it for decades ;)

Removed.

> > @@ -75,8 +75,8 @@ struct ssl_stream {
>      struct ssl_handle *handle;
>      char type;
>      char iseos;
> -    int (*read) ();
> -    void (*close) ();
> +    int (*read) (void *, unsigned char *, int);
> +    void (*close) (void *, int, int *);
> 
> These functions are a bit problematic because they can either be stdio
> functions or unix functions that operat on fd. Specifying `FILE *`
> gives the wrong impression that they're meant for stdio only.

Good spot. I opted to get rid of the io handle struct instead. I think
this is the cleanest solution. I also dropped unnecessary casts.

I did not amend the previous commits to ease reviewing those changes.
This should probably be cleaned up before merging.

Also this is now untested code. I will apply to my local build later.

N-R-K commented on file.c:
```diff
@@ -401,7 +401,7 @@ examineFile(char *path, URLFile *uf)
 	    if ((fp = lessopen_stream(path))) {
 		UFclose(uf);
-		uf->stream = newFileStream(fp, (void (*)())pclose);
+		uf->stream = newFileStream(fp);
```
I haven't carefully checked but won't this lead to trying to close a `FILE*` acquired via `popen` with `fclose`? The manpage states that file streams opened with `popen` must use `pclose` rather than `fclose`:

> The return value from popen() is a normal standard I/O stream in  all
> respects  save  that  it  must  be  closed  with pclose() rather than
> fclose(3)

Also if you can separate the `io_file_handle` commit from the adding `(void)` changes that'd be a bit easier to review.

rkta commented:
On Fri, Dec 22, 2023 at 04:19:14AM -0800, NRK wrote:
> @N-R-K commented on this pull request.
> 
> 
> 
> > @@ -401,7 +401,7 @@ examineFile(char *path, URLFile *uf)
>  		return;
>  	    if ((fp = lessopen_stream(path))) {
>  		UFclose(uf);
> -		uf->stream = newFileStream(fp, (void (*)())pclose);
> +		uf->stream = newFileStream(fp);
> 
> I haven't carefully checked but won't this lead to trying to close a `FILE*` acquired via `popen` with `fclose`? The manpage states that file streams opened with `popen` must use `pclose` rather than `fclose`:
> 
> > The return value from popen() is a normal standard I/O stream in  all
> > respects  save  that  it  must  be  closed  with pclose() rather than
> > fclose(3)

Dammit... Will undo.

> 
> Also if you can separate the `io_file_handle` commit from the adding
> `(void)` changes that'd be a bit easier to review.

Sure, will do.

rkta committed:
Get rid of io_file_handle

This also makes file_close obsolete.

rkta committed:
Remove unused argument check

newFileStream is never called with NULL as second argument, drop the
check and make the argument necessary.

rkta committed:
Remove unnecessary casts for close functions

rkta committed:
Align function signature with standard functions

Using the same signature as fclose and pclose let's us use all functions
without casting.

rkta committed:
Enable warnings for modern C
